Virtual applications docs #441 (#442)
enonic/doc-xp: docs/apps/virtual-applications.adoc

= Virtual Applications
:toc: right
:imagesdir: ../images

image:xp-7110.svg[XP 7.11.0,opts=inline]

Virtual Application allows to extend/override schemas and descriptors from usual bundled application or create separated containers for it. Applications, schemas and descriptors data is stored in nodes in separated `system.app` repository. Use <<../api/lib-app, lib-app>> to manipulate virtual applications.

=== Structure
Virtual application node structure follows bundled applications site's file structure:
[source,files]
----
virtual-app/
    site/
      content-types/
      layouts/
      mixins/
      pages/
      parts/
      x-data/
      site.xml
      styles.xml

----
Use <<../api/lib-schema, lib-schema>> to manipulate schemas and descriptors of virtual apps.

=== Application Mode

> Depends on app config, xp can work with application resources in different modes:

Reviewer: ```suggestion
Depending on app config, XP can work with application resources in different modes:
```

VIRTUAL::

>     If specified resource is presented in both virtual and bundled app, then virtual will be used.

Reviewer: ```suggestion
    If a specified resource is present in both a virtual and a bundled app, then the virtual one will be used.
```

BUNDLED::

>     Resources only from bundled applications are used, virtual apps are disabled and its resources are ignored.

Reviewer: ```suggestion
    Only resources from bundled applications are used, virtual apps are disabled and their resources are ignored.
```

AUGMENTED::

>     If specified resource is presented in both virtual and bundled app, then virtual will be used.

Reviewer: This is exactly the same as description for "VIRTUAL", and doesn't seem right.

> Check the <<../deployment/config#application, documentation>> to find how config the mode.

Reviewer: ```suggestion
Check the <<../deployment/config#application, documentation>> to find out how to configure each mode.
```

= Virtual Applications
:toc: right
:imagesdir: ../images

image:xp-7110.svg[XP 7.11.0,opts=inline]

Virtual Applications allow to extend/override schemas and descriptors from usual bundled application or create a standalone application container. Applications, schemas and descriptors data are stored in nodes in separate `system.app` repository. Use <<../api/lib-app, lib-app>> to manipulate virtual applications.

=== Structure
Node structure of a virtual application follows file structure inside the `site` folder of a bundled application:
[source,files]
----
virtual-app/
    site/
      content-types/
      layouts/
      mixins/
      pages/
      parts/
      x-data/
      site.xml
      styles.xml

----
Use <<../api/lib-schema, lib-schema>> to manipulate schemas and descriptors of virtual apps.

=== Application Mode
Depending on app data and config, the application can work with its resources in different ways:

> VIRTUAL::
> Shows that only virtual application is presented in the system and support of virtual apps is enabled.
> BUNDLED::
> Shows one of the next cases:
> - Only bundled application is presented in the system.
> - Both bundled and virtual applications are presented, but bundled resources override virtual.
> - Both bundled and virtual applications are presented, but virtual app support is disabled.
> AUGMENTED::
> Both bundled and virtual applications are presented in the system and virtual resources override bundled.
> 
> Check the <<../deployment/config#application, documentation>> to find out how to configure application behaviour.

Reviewer: ```suggestion
VIRTUAL::
Shows that only virtual application is present in the system and support of virtual apps is enabled.
BUNDLED::
Can be one of the next cases:
- Only bundled application is present in the system.
- Both bundled and virtual applications are present, but bundled resources are configured to override virtual.
- Both bundled and virtual applications are present, but virtual app support is disabled via config.
AUGMENTED::
Both bundled and virtual applications are present in the system and virtual resources override bundled.

Check the <<../deployment/config#application, documentation>> to find out how to configure application behaviour.
```
